### ISSUE TITLE
GLFW window leaves stuck process after closing on MacOS

### ISSUE THREAD

**Issue by jkrumbiegel** [glfw/glfw#1766]
I'll copy here a comment from @dcabecinhas from issue https://github.com/glfw/glfw/issues/1412 which @elmindreda asked be put into a new issue.

For me, this problem also causes weird behavior when opening new GLFW windows after closing old ones. Each new window appears slightly moved to the right and down, which is the OS trying not to overlap multiple windows (even though the old ones aren't there anymore). So clearly MacOS still thinks that a closed window is still around, which must be because GLFW is not removing it correctly.

I really hope this can be fixed as it reduces the usability of `Makie.jl` on Macs quite a bit. I think graphics resources are also not released correctly because of this which reduces battery capacity quickly.

**Original comment below:**

--------------------------

A window artefact remains in the macOS dock and a hung process shows in the Activity Monitor.app until the main program exists.

The call to GLFW.Terminate() should finalize all GLFW code and windows but that is not happening. The Makie linked issues are precisely because people close the GLFW window, the window disappears from the screen but remain in the dock and as a hung process until the julia process exits.

Having to exit julia to completely finish GLFW is especially troublesome as it clears JIT compiled code and is generally avoided.

System version information:

GLFW version: 3.3.2 Cocoa NSGL EGL OSMesa dynamic
macOS 10.15.6

To reproduce:

```c
#include <stdio.h>
#include <unistd.h>
#include <GLFW/glfw3.h>

int main() {
        glfwInit();
        GLFWwindow* win = glfwCreateWindow(400, 400, "Window", NULL, NULL);
        while (!glfwWindowShouldClose(win)) {
                glfwPollEvents();
        }
        glfwDestroyWindow(win);
        printf("window closed\n");

        glfwTerminate();

        printf("After glfwTerminate()\n");
        sleep(20);
}
```

Notice that a dock icon and a hung process still remain even after After glfwTerminate() is printed to the screen. Further calls to glfwPollEvents() after glfwTerminate() have no effect.

**Comment by jkrumbiegel** [comment]
Any news on this? I can try to help with debugging if someone can tell me where to look approximately? I'm not very proficient with C but maybe it's enough, as long as I have an idea what's involved here.

**Comment by finchco** [comment]
I'm having this same issue and the workaround of adding an extra call to glfwPollEvents() before glfwTerminate() is not working for me.

Version 3.3.3 installed via homebrew. macOS 11.2.2

**Comment by jkrumbiegel** [comment]
Yeah I'd also still love to get this fixed somehow. Sometimes I'm opening and closing a window 20 times until it finally disappears and doesn't remain there in an unresponsive state, and even then it doesn't close correctly. I've looked around in the source code trying to debug but no luck, I assume the error still happens around here? https://github.com/glfw/glfw/blob/master/src/cocoa_window.m#L975-L977

**Comment by SimonDanisch** [comment]
Is there anything we can do to help fixing this?

**Comment by jkrumbiegel** [comment]
I would like to bump this issue again. Is there anything we can do to help resolve it as end users? I understand if the developers maybe don't have access to Apple hardware, I'd be willing to assist in trying debugging steps on my machine. In our data visualization software Makie.jl, this specific issue sadly decreases usability a lot because we're opening GLFW windows in a live REPL session which ideally shouldn't be closed at all (to avoid startup latency). So having these dead unresponsive windows stick around after trying to close them can only be fixed by killing the main julia process right now.

**Comment by ws909** [comment]
In Cocoa, `NSApplication` is the class responsible for interacting with the window server, display server, etc. All main Cocoa application components will be alive for the lifetime of an `NSApplication` instance. NSApplication registers the running executable as an interactive GUI application with MacOS, at `NSApplication run:`. The menu bar, Dock tile, and Dock tile context menu are handled by the application itself, but to bring any of these up, the user interacts with an external program (WindowServer or Finder, I believe), which then sends a message to the application. Unless the application's main `CFRunLoop` is waiting/running, the application will not perform these actions, and will neither respond to this external program, which makes that process mark the application as unresponsive. `NSApplication terminate:` is required to quit the application, but this does not transfer control back to the caller; instead it terminates the running process with `quit(0)`. As such, GLFW can not call this method (which it doesn't), and a new `NSApplication` cannot be created afterwards (obviously because the program is shut down).

GLFW currently only implements `NSAppDelegate`. It might be possible to resolve this issue by subclassing `NSApplication`, but I doubt so. Likely, something more unconventional is required.

**Comment by jkrumbiegel** [comment]
Thank you for the explanation, it helped me realize that the problems we were seeing on Mac were two separate things. So even though the stuck process icon in the dock is annoying, it's not as bad as the stuck windows that I thought were related, but seem not to be. I will not attempt to fix the stuck process issue at this time in glfw, as it's not pressing enough to prioritize over other issues, and because it sounds like a solution would be nontrivial if it's possible at all with our REPL process setup.

**Comment by ws909** [comment]
Always happy to help. :) 
The «stuck process problem» is best fixed by never calling `glfwTerminate` until the process actually is supposed to quit. Rather, use a custom flag instead of `glfwWindowShouldClose` for the main event loop, and call `glfwPollEvents` or `glfwWaitEvents` for the entire lifetime of the process (no windows are necessary). Handle windows seperate from event polling/waiting.

**Comment by ws909** [comment]
Some kind of pseudo-code:
```
typedef GLFWWindow Window;

const List<Window> windows;
boolean run = true;

Window newWindow() {
    window = glfwCreateWindow(800, 540, "Window", null, null);
    windows.add(window);
    return window;
}

void main() {
    glfwInit();

    while (run) {
        glfwWaitEvents();
        
        for (window : windows) {
            if (glfwWindowShouldClose(window) {
                glfwDestroyWindow(window);
                windows.remove(window);
                continue;
            }
            additionalLogic(window);
            render(window);
        }
    }
    
    for (window : windows) {
        glfwDestroyWindow(window);
    }

    glfwTerminate();
}
```

**Comment by ws909** [comment]
You can use this to hide the Dock icon and menu bar when there's no windows open:
`NSApp.activationPolicy = NSApplicationActivationPolicyAccessory;`

**Comment by ws909** [comment]
Potential fixes for the issue:

1. Set the activation policy to `Accessory` in `glfwInit`. Do nothing in `glfwTerminate`.
2. Set the activation policy to `Regular` in `glfwInit`, and `Prohibited` in `glfwTerminate`.
3. Set the activation policy to `Accessory` in `glfwTerminate`[^1].

Both of these will cause issues for applications specifying the [activation policy](https://developer.apple.com/documentation/appkit/nsapplicationactivationpolicy?language=objc) themselves.

1. Has been tested, and does not solve the issue. The dock icon remains hidden even when windows are open.

2. Works as expected, except that the second window is shown onscreen for 0.5 seconds, before losing focus, dropping behind other windows on the desktop. Might be a result of my testing environment. App focus is transferred back to Xcode when the first window is closed (or first time `glfwTerminate` is called), in this setup.

3. Everything works as expected. The issue with solution 2 is not occurring. 

[^1]: GLFW already sets the activation policy to `Regular` in `glfwInit` if the application is unbundled. To the best of my knowledge, the GLFW test targets are unbundled.

It's easier to test/debug with this:
```c
int main() {
    printf("[0] Before glfwInit\n");
    glfwInit();
    printf("[1] After glfwInit\n");
    GLFWwindow* win = glfwCreateWindow(400, 400, "Window 1", NULL, NULL);
    while (!glfwWindowShouldClose(win)) {
            glfwPollEvents();
    }
    glfwDestroyWindow(win);
    printf("[2] window closed\n");

    glfwTerminate();

    printf("[3] After glfwTerminate()\n");
    sleep(5);

    printf("[4] Before glfwInit\n");
    glfwInit();
    printf("[5] After glfwInit\n");
    win = glfwCreateWindow(400, 400, "Window 2", NULL, NULL);
    while (!glfwWindowShouldClose(win)) {
            glfwPollEvents();
    }
    glfwDestroyWindow(win);
    printf("[6] window closed\n");

    glfwTerminate();

    printf("[7] After glfwTerminate()\n");
    sleep(5);
    printf("[8] End\n");
}
```

**Comment by jkrumbiegel** [comment]
Oh that would be nice, I'll have to check how to access that from GLFW.jl if it's possible at all.

**Comment by ws909** [comment]
You'll have to use Objective-C for it. I can create a PR with solution 3, and set it to `Regular` in `glfwInit` regardless of the app being bundled or not, but I'll have to dig up why GLFW currently only does it for unbundled applications, or wait for a comment from one of the core maintainers. It's most likely because bundled applications ship with a `.plist` file that contain an entry for the activation policy. This does also risk breaking existing applications, but due to only happening in `glfwInit` and `glfwTerminate`, is probably a minor problem, or not one at all.

**Comment by ws909** [comment]
A workaround for that might be to also check if the application is unbundled, and only apply this "fix" to unbundled applications, or save the currently set policy at the end of `glfwInit` for the first call, and then set this policy each time for subsequent calls to `glfwInit` after `glfwTerminate` has been applied.

**Comment by jkrumbiegel** [comment]
> You'll have to use Objective-C or Swift for it

I'm not sure how exactly GLFW.jl works but I assume some fixed version is compiled into a shared library, so I cannot set any specific build flags etc. And I'd have to check about Cocoa bindings as well.

**Comment by ws909** [comment]
Just apply it on top of GLFW. `NSApp`/`NSApplication.sharedApplication` is a global variable. Set it after `glfwInit`.
